components/sensor/power_ina219/power_ina219.go

```diff
@@ -0,0 +1,229 @@
+	powerRegister        = 0x03
+	currentRegister      = 0x04
+	calibrationRegister  = 0x05
+)
+
+// AttrConfig is used for converting config attributes.
+type AttrConfig struct {
+	Board   string `json:"board"`
+	I2CBus  string `json:"i2c_bus"`
+	I2cAddr int    `json:"i2c_addr,omitempty"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *AttrConfig) Validate(path string) ([]string, error) {
+	var deps []string
+	if len(config.Board) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "board")
+	}
+	deps = append(deps, config.Board)
+	if len(config.I2CBus) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "i2c_bus")
+	}
+	return deps, nil
+}
+
+func init() {
+	registry.RegisterComponent(
+		sensor.Subtype,
+		modelname,
+		registry.Component{Constructor: func(
+			ctx context.Context,
+			deps registry.Dependencies,
+			config config.Component,
+			logger golog.Logger,
+		) (interface{}, error) {
+			attr, ok := config.ConvertedAttributes.(*AttrConfig)
+			if !ok {
+				return nil, rdkutils.NewUnexpectedTypeError(AttrConfig{}, config.ConvertedAttributes)
+			}
+			return newSensor(deps, attr, logger)
+		}})
+
+	config.RegisterComponentAttributeMapConverter(sensor.Subtype, modelname,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		}, &AttrConfig{})
+}
+
+func newSensor(
+	deps registry.Dependencies,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
+		return nil, fmt.Errorf("ina219 init: failed to find board: %w", err)
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, fmt.Errorf("board %s is not local", attr.Board)
+	}
+	i2cbus, ok := localB.I2CByName(attr.I2CBus)
+	if !ok {
+		return nil, fmt.Errorf("ina219 init: failed to find i2c bus %s", attr.I2CBus)
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Infof("using i2c address : %d", defaultI2Caddr)
+	}
+
+	s := &ina219{
+		logger: logger,
+		bus:    i2cbus,
+		addr:   byte(addr),
+	}
+
+	err = s.calibrate()
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power.
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	addr       byte
+	currentLSB int64
+	powerLSB   int64
+	cal        uint16
+}
+
+type powerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate() error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
```

Comment:
I trust these calculations are right; I don't see them in the datasheet but I skipped most of it. Worth putting in a comment to explain them? Maybe not.